esp/services/ws_fs/ws_fsService.cpp

```diff
@@ -3076,34 +2981,50 @@ void CFileSprayEx::addDropZoneFile(IEspContext& context, IDirectoryIterator* di,
     timestr.appendf("%04d-%02d-%02d %02d:%02d:%02d", y,m,d,h,min,sec);
     aFile->setModifiedtime(timestr.str());
     aFile->setFilesize(di->getFileSize());
-    aFile->setServer(server);
+    if (version >= 1.23)
+        aFile->setServer(server);
     files.append(*aFile.getLink());
 }
 
-void CFileSprayEx::searchDropZoneFiles(IEspContext& context, const char* server, const char* dir, const char* nameFilter, IArrayOf<IEspPhysicalFileStruct>& files, unsigned& filesFound)
+bool CFileSprayEx::searchDropZoneFiles(IEspContext& context, const char* dropZoneName, const char* server,
+    const char* dir, const char* relDir, const char* nameFilter, IArrayOf<IConstPhysicalFileStruct>& files, unsigned& filesFound)
 {
+    if (getDropZoneScopePermissions(context, dropZoneName, dir, server) < SecAccess_Read)
+        return false;
+
     RemoteFilename rfn;
     SocketEndpoint ep(server);
     rfn.setPath(ep, dir);
     Owned<IFile> f = createIFile(rfn);
     if(f->isDirectory()!=fileBool::foundYes)
-        throw MakeStringException(ECLWATCH_INVALID_DIRECTORY, "%s is not a directory.", dir);
+        throw makeStringExceptionV(ECLWATCH_INVALID_DIRECTORY, "%s is not a directory.", dir);
 
-    const char pathSep = getPathSepChar(dir);
-    Owned<IDirectoryIterator> di = f->directoryFiles(nameFilter, true, true);
+    Owned<IDirectoryIterator> di = f->directoryFiles(nullptr, false, true);
```

Comment:
What are the performance implications of this change?  Previously the filter was provided to the iterator, now it is post-filtered.  If the source is remote blob storage that may end up with a significantly higher number of scan operations.
@jakesmith As a slightly separate issue, should we be monitoring the total file costs for esp and reporting it somewhere - in case we have invisible costs that are not being recognised. HPCC-29099 created to discuss.

Reply:
The main difference here I think, is that previously it perform a recursive iteration of sub dirs, seeing if there were any filter matches all the way down.
With this change, because it is scope checking, that can be pruned (no point recursing down subdirs where user has no access).
So that potential saves overhead accessing storage that it would have done previously.

I don't think the implementation (of the directory iterator) means there would be any practical different in performance of passing filter in, vs post filtering, as it is effectively post-filtering in the implementation itself.

However, other/future iterations could be different, and I don't recall why nameFilter isn't being passed to the iterator, vs post-filtering, @wangkx  ?

Reply:
>I don't think the implementation (of the directory iterator) means there would be any practical different in performance of passing filter in, vs post filtering, as it is effectively post-filtering in the implementation itself.

Not quite true, if the file is a link (or entry->d_type==DT_UNKNOWN), it will perform an additional stat() on call on those files.
It would be better to pass the nameFilter to the iterator unless there was a good reason not to.

Reply:
This is what I remember why nameFilter isn't being passed to the iterator. Let's say we want to find out the files which file name begins with 'f'. One file 'fxyz' is in the subfolder 'dxyz'. If we use the nameFilter 'f*', the  subfolder 'dxyz' will be skipped and we will not find out the file 'fxyz'.

Reply:
ok yes, that makes sense, because the filter is also applied to dirs, *but* sub=true recursed down all directories (unfiltered). Slightly odd semantics.

Pre-PR: (sub=true, includedirs=true + nameFilter) recursed down all dirs, applying filter to files - costly both on FS and Scope checking.
Post-PR: (sub=false, includedirs=true) returns all at a given file level, post filters filenames, continues down directories if has permissions.

We could change directoryFiles so it had more control, e.g. so that it had options such that it could filter on filenames and return all directories (e.g. by having a different filter for directories perhaps).
But not sure worth it in general, and definitely not at this juncture.

I think it is okay as it stands.